nifi-nar-bundles/nifi-smb-bundle/nifi-smb-processors/src/main/java/org/apache/nifi/processors/smb/ListSmb.java

```diff
@@ -0,0 +1,342 @@
+            ListedEntityTracker.INITIAL_LISTING_TARGET
+    ));
+
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return PROPERTIES;
+    }
+
+    @Override
+    protected Map<String, String> createAttributes(SmbListableEntity entity, ProcessContext context) {
+        final Map<String, String> attributes = new TreeMap<>();
+        attributes.put("filename", entity.getName());
+        attributes.put("shortname", entity.getShortName());
+        attributes.put("path", entity.getPath());
+        attributes.put("absolute.path", getPath(context) + entity.getPathWithName());
+        attributes.put("identifier", entity.getIdentifier());
+        attributes.put("timestamp",
+                ISO_DATE_TIME.format(LocalDateTime.ofEpochSecond(entity.getTimestamp(), 0, ZoneOffset.UTC)));
+        attributes.put("creationTime",
+                ISO_DATE_TIME.format(LocalDateTime.ofEpochSecond(entity.getCreationTime(), 0, ZoneOffset.UTC)));
+        attributes.put("lastAccessedTime",
+                ISO_DATE_TIME.format(LocalDateTime.ofEpochSecond(entity.getLastAccessTime(), 0, ZoneOffset.UTC)));
+        attributes.put("changeTime",
+                ISO_DATE_TIME.format(LocalDateTime.ofEpochSecond(entity.getChangeTime(), 0, ZoneOffset.UTC)));
+        attributes.put("size", String.valueOf(entity.getSize()));
+        attributes.put("allocationSize", String.valueOf(entity.getAllocationSize()));
+        return unmodifiableMap(attributes);
+    }
+
+    @Override
+    protected String getPath(ProcessContext context) {
+        final SmbClientProviderService connectionPoolService =
+                context.getProperty(SMB_CONNECTION_POOL_SERVICE).asControllerService(SmbClientProviderService.class);
+        final URI serviceLocation = connectionPoolService.getServiceLocation();
+        final String directory = getDirectory(context);
+        return String.format("%s:\\%s", serviceLocation.toString(), directory.isEmpty() ? "" : directory + "\\");
+    }
+
+    @Override
+    protected List<SmbListableEntity> performListing(ProcessContext context, Long minimumTimestampOrNull,
+            ListingMode listingMode) throws IOException {
+
+        final Predicate<SmbListableEntity> fileFilter =
+                createFileFilter(context, minimumTimestampOrNull);
+
+        try (Stream<SmbListableEntity> listing = performListing(context)) {
+            final Iterator<SmbListableEntity> iterator = listing.iterator();
+            final List<SmbListableEntity> result = new LinkedList<>();
+            while (iterator.hasNext()) {
+                if (!isExecutionScheduled(listingMode)) {
+                    return emptyList();
+                }
+                final SmbListableEntity entity = iterator.next();
+                if (fileFilter.test(entity)) {
+                    result.add(entity);
+                }
+            }
+            return result;
```

Comment:
Having a `Stream` at hand, could not foreach + collect be used?

Reply:
Please note that a processor can be stopped during listing.